test/functional/wallet_conflicts.py

```diff
@@ -123,5 +129,148 @@ def run_test(self):
         assert_equal(former_conflicted["confirmations"], 1)
         assert_equal(former_conflicted["blockheight"], 217)
 
+    def test_mempool_conflict(self):
+        self.nodes[0].createwallet("alice")
+        alice = self.nodes[0].get_wallet_rpc("alice")
+
+        self.nodes[1].createwallet("bob")
+        bob = self.nodes[1].get_wallet_rpc("bob")
+
+        self.generatetoaddress(self.nodes[0], COINBASE_MATURITY + 3, alice.getnewaddress())
+
+        self.log.info("Test a scenario where a transaction has a mempool conflict")
+
+        unspents = [{"txid" : element["txid"], "vout" : element["vout"]} for element in alice.listunspent()]
+        assert_equal(len(unspents), 3)
+
+        raw_tx = alice.createrawtransaction(inputs=[unspents[0], unspents[1]], outputs=[{bob.getnewaddress() : 49.9999}])
+        tx1 = alice.signrawtransactionwithwallet(raw_tx)['hex']
+
+        raw_tx = alice.createrawtransaction(inputs=[unspents[1], unspents[2]], outputs=[{bob.getnewaddress() : 49.99}])
+        tx2 = alice.signrawtransactionwithwallet(raw_tx)['hex']
+
+        raw_tx = alice.createrawtransaction(inputs=[unspents[2]], outputs=[{bob.getnewaddress() : 24.9899}])
+        tx3 = alice.signrawtransactionwithwallet(raw_tx)['hex']
+
+        tx1 = alice.sendrawtransaction(tx1)
+
+        assert_equal(alice.listunspent()[0]["txid"], unspents[2]["txid"])
+        assert_equal(alice.getbalance(), 25)
+
+        tx2 = alice.sendrawtransaction(tx2)
+
+        # Check that the 0th unspent is now available because the transaction spending it has been replaced in the mempool
+        assert_equal(alice.listunspent()[0]["txid"], unspents[0]["txid"])
+        assert_equal(alice.getbalance(), 25)
+
+        assert_equal(alice.gettransaction(tx1)["mempoolconflicts"], [tx2])
+
+        self.log.info("Test scenario where a mempool conflict is removed")
+
+        tx3 = alice.sendrawtransaction(tx3)
+
+        assert_equal(alice.gettransaction(tx1)["mempoolconflicts"], [])
+
+        # now all of alice's outputs should be considered spent
```

Comment:
In 003efbbe45079c4416810a025b2bc372559dff15 "test: Add tests for wallet mempool conflicts":
Since `tx2` replaced `tx1`, why would `unspents[0]` be considered spent here?

Reply:
Because `tx3` replaced `tx2`, so `tx1` is now `TxStateInactive` instead of `TxStateMempoolConflicted`, making `unspents[0]` spent.

Reply:
I have added a comment before all of the tests in this file explaining how these tests create conflicts.